opm/simulators/linalg/mixed/prec.c

```diff
@@ -0,0 +1,627 @@
+    bsr_init(P->D, nrows, nrows, b);
+    bsr_init(P->U, nrows, (nnz-nrows)/2, b);
+
+    // Sparsity of L,D, and U components of matrix A.
+    // Sparsity of L is expressed for its transpose
+    // which is identical to the sparsity of U due to
+    // the assumption of structural symmetry.
+    nnz=0;
+    for (int i=0;i<nrows;i++)
+    {
+        L->rowptr[i]=nnz;
+        D->rowptr[i]=i;
+        U->rowptr[i]=nnz;
+        for (int k=A->rowptr[i];k<A->rowptr[i+1];k++)
+        {
+            int j=A->colidx[k];
+            if(j>i)
+            {
+                L->colidx[nnz]=j;
+                U->colidx[nnz]=j;
+                nnz++;
+            }
+        }
+        D->colidx[i]=i;
+    }
+    L->rowptr[nrows]=nnz;
+    D->rowptr[nrows]=nrows;
+    U->rowptr[nrows]=nnz;
+
+    // replace with asserts?
+    L->nnz=nnz;
+    D->nnz=nrows;
+    U->nnz=nnz;
+
+
+    // offsets for off-diagonal updates
+    int count;
+    count = prec_analyze(U,NULL);
+    P->offsets = malloc(3*(count+1)*sizeof(int));
+    assert(P->offsets);
+    count = prec_analyze(U,P->offsets);
+    P->offsets[count][0]=U->nnz;
+    P->noffsets=count;
+}
+
+void mat3_matmul(double *C, const double *A, const double *B)
+{
+    // assume 3x3 column-major matrices
+    // account for possiblity of C and A referring to same memory location
+    double M[9];
+    for(int k=0;k<9;k++) M[k]=0.0;
+    for(int j=0;j<3;j++)
+    {
+        double *m_j = M+3*j;        // j-th column of M
+        for(int k=0;k<3;k++)
+        {
+            double b_kj = B[k+3*j]; // kj-th element of B
+            double const *a_k = A+3*k;    // k-th column of A
+            for(int i=0;i<3;i++) m_j[i] += a_k[i]*b_kj; // |m_j> += |a_k> * b_kj;
+        }
+    }
+    for(int k=0;k<9;k++) C[k]=M[k];
+}
+
+void mat3_matfms(double *C, const double *A, const double *B)
+{
+    // assume 3x3 column-major matrices
+    // account for possiblity of C and A referring to same memory location
+    double M[9];
+    for(int k=0;k<9;k++) M[k]=0.0;
+    for(int j=0;j<3;j++)
+    {
+        double *m_j = M+3*j;        // j-th column of M
+        for(int k=0;k<3;k++)
+        {
+            double b_kj = B[k+3*j]; // kj-th element of B
+            double const *a_k = A+3*k;    // k-th column of A
+            for(int i=0;i<3;i++) m_j[i] += a_k[i]*b_kj; // |m_j> += |a_k> * b_kj;
+        }
+    }
+    for(int k=0;k<9;k++) C[k]-=M[k];
+}
+
+
+void mat3_inv(double *invA, const double *A)
+{
+    double M[9];
+    M[0] = +(A[4]*A[8]-A[5]*A[7]);
+    M[3] = -(A[3]*A[8]-A[5]*A[6]);
+    M[6] = +(A[3]*A[7]-A[4]*A[6]);
+    M[1] = -(A[1]*A[8]-A[2]*A[7]);
+    M[4] = +(A[0]*A[8]-A[2]*A[6]);
+    M[7] = -(A[0]*A[7]-A[1]*A[6]);
+    M[2] = +(A[1]*A[5]-A[2]*A[4]);
+    M[5] = -(A[0]*A[5]-A[2]*A[3]);
+    M[8] = +(A[0]*A[4]-A[1]*A[3]);
+
+    double detA = A[0]*M[0]+A[1]*M[3]+A[2]*M[6];
+    for(int k=0;k<9;k++) invA[k]=M[k]/detA;
+}
+
+inline void vec_copy9(double *y, double const *x)
```

Comment:
Besides all this nitpicking above there is a serious issue. The code does not link because of inline functions when I compile with `-O0`:
```
[100%] Linking CXX executable bin/flow
/usr/bin/ld: lib/libopmsimulators.so.2026.04: undefined reference to `vec_copy9'
collect2: error: ld returned 1 exit status
make[3]: *** [CMakeFiles/flow.dir/build.make:233: bin/flow] Fehler 1
make[2]: *** [CMakeFiles/Makefile2:6973: CMakeFiles/flow.dir/all] Fehler 2
make[1]: *** [CMakeFiles/Makefile2:6980: CMakeFiles/flow.dir/rule] Fehler 2
make: *** [Makefile:3195: flow] error 2
```

```suggestion
void vec_copy9(double *y, double const *x)
```
or find another way to include the symbol. The inline is only a hint. Is it really needed?  https://stackoverflow.com/questions/19068705/undefined-reference-when-calling-inline-function

The other inline functions here have no problem. That might be because they are actually not inlined?

Reply:
> Besides all this nitpicking above there is a serious issue. The code does not link because of inline functions when I compile with `-O0`

So far, I have not been able to reproduced the issue on my end. Do you mind trying
```c
static inline void vec_copy9(double *y, double const *x)
```
It should limit the scope to the file where the function is defined. I'll keep trying to reproduce the issue.